pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1882,6 +2090,29 @@ func (og *operationGenerator) nodeExpandVolume(
 	return true, nil
 }
 
+func permitNodeExpansion(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) bool {
+	pvcStatusCap := pvc.Status.Capacity[v1.ResourceStorage]
+	pvSpecCap := pv.Spec.Capacity[v1.ResourceStorage]
+	// if pvc.Status.Cap is >= pv.Spec.Cap then volume is already expanded
+	if pvcStatusCap.Cmp(pvSpecCap) >= 0 {
+		return false
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.RecoverVolumeExpansionFailure) {
+		resizeStatus := pvc.Status.ResizeStatus
+		// if resizestatus is nil or NodeExpansionInProgress or NodeExpansionPending then we should allow volume expansion on
+		// the node to proceed. We are making an exception for resizeStatus being nil because it will support use cases where
+		// resizeStatus may not be set (old control-plane expansion controller etc).
+		if resizeStatus == nil || *resizeStatus == v1.PersistentVolumeClaimNodeExpansionPending || *resizeStatus == v1.PersistentVolumeClaimNodeExpansionInProgress {
+			return true
+		} else {
+			klog.Infof("volume %s/%s can not be expanded because resizeStaus is: %s", pvc.Namespace, pvc.Name, *resizeStatus)
```

Comment:
How often can this be logged?

Reply:
I would think - this should be somewhat rare and only when expansion is failing on the node or somehow pvc is in weird state and asking controller to resync it.